renderdoc/replay/entry_points.cpp

```diff
@@ -890,6 +1203,208 @@ extern "C" RENDERDOC_API void RENDERDOC_CC RENDERDOC_StartAndroidRemoteServer(co
       "shell am start -n org.renderdoc.renderdoccmd/.Loader -e renderdoccmd remoteserver");
 }
 
+extern "C" RENDERDOC_API bool RENDERDOC_CC RENDERDOC_CheckAndroidLayer(const char *host,
+                                                                       const char *exe)
+{
+  string packageName(basename(string(exe)));
+
+  int index = 0;
+  std::string deviceID;
+  Android::extractDeviceIDAndIndex(host, index, deviceID);
+
+  // Find the path to package
+  string pkgPath = trim(adbExecCommand(deviceID, "shell pm path " + packageName).strStdout);
+  pkgPath.erase(pkgPath.begin(), pkgPath.begin() + strlen("package:"));
+  pkgPath.erase(pkgPath.end() - strlen("base.apk"), pkgPath.end());
+  pkgPath += "lib";
+
+  string layerName = "libVkLayer_GLES_RenderDoc.so";
+
+  // First, see if the application contains the layer
+  if(SearchForAndroidLayer(deviceID, pkgPath, layerName))
+    return true;
+
+  // Next, check a debug location only usable by rooted devices
+  if(SearchForAndroidLayer(deviceID, "/data/local/debug/vulkan", layerName))
+    return true;
+
+  // TODO: Add any future layer locations
+
+  RDCWARN("No RenderDoc layer for Vulkan was found");
+  return false;
+}
+
+string DetermineInstalledABI(const string &deviceID, const string &packageName)
+{
+  RDCLOG("Checking installed ABI for %s", packageName.c_str());
+  string abi;
+
+  string dump = adbExecCommand(deviceID, "shell pm dump " + packageName).strStdout;
+  if(dump.empty())
+    RDCERR("Unable to pm dump %s", packageName.c_str());
+
+  // Walk through the output and look for primaryCpuAbi
+  std::istringstream contents(dump);
+  string line;
+  string prefix("primaryCpuAbi=");
+  while(std::getline(contents, line))
+  {
+    line = trim(line);
+    if(line.compare(0, prefix.size(), prefix) == 0)
+    {
+      // Extract the abi
+      abi = line.substr(line.find_last_of("=") + 1);
+      RDCLOG("primaryCpuAbi found: %s", abi.c_str());
+      break;
+    }
+  }
+
+  if(abi.empty())
+    RDCERR("Unable to determine installed abi for: %s", packageName.c_str());
+
+  return abi;
+}
+
+string FindAndroidLayer(const string &abi, const string &layerName)
+{
+  string layer;
+
+  // Check known paths for RenderDoc layer
+  string exePath;
+  FileIO::GetExecutableFilename(exePath);
+  string exeDir = dirname(FileIO::GetFullPathname(exePath));
+
+  std::vector<std::string> paths;
+
+#if defined(RENDERDOC_LAYER_PATH)
+  string customPath(RENDERDOC_LAYER_PATH);
+  RDCLOG("Custom layer path: %s", customPath.c_str());
+
+  if(FileIO::IsRelativePath(customPath))
+    customPath = exeDir + "/" + customPath;
+
+  // Check to see if layer name was included in custom path
+  if(!endswith(customPath, layerName))
+  {
+    if(customPath.back() != '/')
+      customPath += "/";
+    customPath += layerName;
+  }
+  paths.push_back(customPath);
+#endif
+
+  // clang-format off
```

Comment:
Yeh clang-format isn't perfect but the alternatives are me nit-picking all the little formatting things manually, or letting it slide completely and ending up with a codebase with mixed space and tabs, inconsistent formatting, etc.

No-one can ever agree on a coding style, but at least as much as possible clang-format takes any arguments away by automating the process :smile:. Plus since it can be so automated it means you can write the code however you feel comfortable with and just apply the format before committing. There's some stuff that I would prefer to do differently but I couldn't represent it in the clang-format file, so I just write my personal style and it gets reformatted when I save.